kbpagesconfig/common.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/keybase/kbfs/libpages/config"
+	"github.com/sergi/go-diff/diffmatchpatch"
+)
+
+type prompter interface {
+	Prompt(string) (string, error)
+	PromptPassword(string) (string, error)
+}
+
+// not go-routine safe!
+type kbpConfigEditor struct {
+	kbpConfigPath     string
+	kbpConfig         *config.V2
+	originalConfigStr string
+	prompter          prompter
+}
+
+func readConfigAndClose(from io.ReadCloser) (
+	cfg config.Config, str string, err error) {
+	defer from.Close()
+	buf := &bytes.Buffer{}
+	if cfg, err = config.ParseConfig(io.TeeReader(from, buf)); err != nil {
+		return nil, "", err
+	}
+	return cfg, buf.String(), nil
+}
+
+func kbpConfigPath(kbpConfigDir string) (path string, err error) {
+	fi, err := os.Stat(kbpConfigDir)
+	if err != nil {
+		return "", fmt.Errorf("stat %q error: %v", kbpConfigDir, err)
+	}
+	if !fi.IsDir() {
+		return "", fmt.Errorf("%q is not a directory", kbpConfigDir)
+	}
+	return filepath.Join(kbpConfigDir, config.DefaultConfigFilename), nil
+}
+
+func promptConfirm(p prompter, text string, defaultYes bool) (confirmed bool, err error) {
+	prompt := "(y/N)"
+	if defaultYes {
+		prompt = "(Y/n)"
+	}
+	input, err := p.Prompt(fmt.Sprintf("%s %s: ", text, prompt))
+	if err != nil {
+		return false, fmt.Errorf("getting confirmation error: %v", err)
+	}
+	switch strings.ToLower(input) {
+	case "y":
+		return true, nil
+	case "n":
+		return false, nil
+	case "":
+		return defaultYes, nil
+	default:
+		return false, errors.New("getting confirmation error")
+	}
+}
+
+func confirmAndWrite(
+	originalConfigStr string,
+	newConfig config.Config,
+	configPath string,
+	p prompter) (err error) {
+	buf := &bytes.Buffer{}
+	if err := newConfig.Encode(buf, true); err != nil {
+		return fmt.Errorf("encoding config error: %v", err)
+	}
+	newConfigStr := buf.String()
+	if newConfigStr == originalConfigStr {
```

Comment:
Is there some guarantee that the json encoder will output struct fields in the same order every time?  I don't see that in the documentation.  Are you sure this is safe?

Reply:
Good question. It does mention that map keys are sorted, but didn't mention anything about object keys that are from structs. It does seem like it's sorted by the order they are defined in the struct, but not being in the doc means it might go away.

Although I think it's safe in this case though, since even if the order change, we'd be showing an annoying diff to user and write the config when it's not necessary. But we'd never skip a write when we should. I can add a test for the struct ordering so that we catch it when this behavior change in the stdlib.